lib/if.c

```diff
@@ -413,13 +413,13 @@ struct interface *if_lookup_by_name_all_vrf(const char *name)
 
 struct interface *if_lookup_by_index_all_vrf(ifindex_t ifindex)
 {
-	struct vrf *vrf;
+	struct vrf *vrf, *vrf_next;
 	struct interface *ifp;
 
 	if (ifindex == IFINDEX_INTERNAL)
 		return NULL;
 
-	RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
+	RB_FOREACH_SAFE (vrf, vrf_id_head, &vrfs_by_id, vrf_next) {
```

Comment:
I am trying to understand how did it lead to core in first place. This is lookup function and here we are looing all the VRF in one go and then doing a lookup in that for that VRF-id. So we do not have any space for any event to delete VRF here.

Reply:
we have 2 pthreads iterating over this structure on shutdown.
1 cleaning it up, the other attempting to use it.

This will probably prevent an immediate crash but not really stop the small window of opportunity here.  At the moment I probably would be fine with this.

Reply:
Thanks Donald. In that case safe walk may not be safe and need locking?

Reply:
digging a bit deeper, this is what the main pthread is doing on a SIGTERM:
```c
	frr_early_fini();
	zebra_dplane_pre_finish();
[...]
	vrf_terminate();
[...]
	/* Indicate that all new dplane work has been enqueued. When that
	 * work is complete, the dataplane will enqueue an event
	 * with the 'finalize' function.
	 */
	zebra_dplane_finish();
```
So we could avoid doing the lookup from the dplane thread once we're notified of the pre-finish... except we don't:
```c
void zebra_dplane_pre_finish(void)
{
	if (IS_ZEBRA_DEBUG_DPLANE)
		zlog_debug("Zebra dataplane pre-fini called");

	zdplane_info.dg_is_shutdown = true;

	/* TODO -- Notify provider(s) of pending shutdown */
}
```
However there is a `bool dplane_is_in_shutdown(void)` that could be used to check whether we should stop looking up VRFs in the dplane provider because everything is being shut down. So short of the actual notification TODO, the fix above and checking the flag in the provider plugin before walking the VRF table should be sufficient for the time being

Reply:
I wonder if there was not a crash when using zebra netns. actually, removing a VRF may lead to a crash on the dataplane working on some jobs of that namespace.
it would be nice to add a finer grain by making a protection vrf by vrf.